File: gj06-backend/api/push.py
import os
from fastapi import APIRouter, HTTPException
from lib.supabase_client import supabase
from schemas import PushSubscription
from dotenv import load_dotenv

load_dotenv()

router = APIRouter()

VAPID_PUBLIC_KEY = os.environ.get("VAPID_PUBLIC_KEY", "")


@router.get("/push/vapid-key")
def get_vapid_key():
    """Frontend fetches this to set up push subscription."""
    if not VAPID_PUBLIC_KEY:
        raise HTTPException(status_code=500, detail="VAPID not configured")
    return {"public_key": VAPID_PUBLIC_KEY}


@router.post("/push/subscribe")
def subscribe(body: PushSubscription):
    """
    Owner's browser/phone calls this after granting notification permission.
    Saves the subscription so we can send push notifications.
    """
    try:
        data = {
            "endpoint": body.endpoint,
            "p256dh":   body.keys.get("p256dh", ""),
            "auth":     body.keys.get("auth", ""),
        }

        # Upsert — update if endpoint already exists
        supabase.table("push_subscriptions").upsert(
            data, on_conflict="endpoint"
        ).execute()

        return {"message": "Subscribed to push notifications"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/push/unsubscribe")
def unsubscribe(endpoint: str):
    """Remove a push subscription."""
    try:
        supabase.table("push_subscriptions").delete().eq(
            "endpoint", endpoint
        ).execute()
        return {"message": "Unsubscribed"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: pages/api/push.js
import { supabaseAdmin } from '../../lib/supabase';

export default async function handler(req, res) {
  // GET — return VAPID public key
  if (req.method === 'GET') {
    return res.status(200).json({ publicKey: process.env.VAPID_PUBLIC_KEY || '' });
  }

  // POST — save push subscription
  if (req.method === 'POST') {
    const { endpoint, p256dh, auth } = req.body;

    if (!endpoint || !p256dh || !auth) {
      return res.status(400).json({ error: 'Missing subscription fields' });
    }

    // Upsert by endpoint so re-subscribing doesn't duplicate
    const { error } = await supabaseAdmin
      .from('push_subscriptions')
      .upsert({ endpoint, p256dh, auth }, { onConflict: 'endpoint' });

    if (error) {
      console.error('Push subscription save error:', error);
      return res.status(500).json({ error: 'Failed to save subscription' });
    }

    return res.status(201).json({ success: true });
  }

  return res.status(405).json({ error: 'Method not allowed' });
}
